// ratemyprofessors-client/dist/config.js
/**
 * Configuration for the RateMyProfessors API client.
 *
 * All client behavior (base URL, timeouts, retries, rate limiting) is driven
 * by {@link RMPClientConfig}. Build one with {@link createConfig},
 * then pass it to {@link RMPClient}.
 */
/** GraphQL endpoint URL. */
export const DEFAULT_BASE_URL = "https://www.ratemyprofessors.com/graphql";
const DEFAULT_USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:129.0) Gecko/20100101 Firefox/129.0";
const defaultHeaders = {
    "User-Agent": DEFAULT_USER_AGENT,
    "Accept-Language": "en-US,en;q=0.5",
};
/**
 * Builds a full config by merging optional overrides onto defaults.
 *
 * @param overrides - Partial config; only provided keys override defaults.
 * @returns A complete {@link RMPClientConfig}.
 */
export function createConfig(overrides = {}) {
    return {
        base_url: DEFAULT_BASE_URL,
        timeout_seconds: 10,
        max_retries: 3,
        rate_limit_per_minute: 60,
        user_agent: DEFAULT_USER_AGENT,
        default_headers: { ...defaultHeaders },
        ...overrides,
    };
}
