// ratemyprofessors-client/dist/http.js
/**
 * HTTP client used by the RMP client: rate limiting, retries, and error mapping.
 *
 * - All requests go through a token-bucket rate limiter (see {@link TokenBucket}).
 * - Failed requests (5xx or network errors) are retried up to config.max_retries.
 * - Non-2xx responses become {@link HttpError}; GraphQL `errors` in the body become {@link RMPAPIError}.
 * - Timeouts use AbortController; call {@link close} to cancel in-flight requests.
 */
import { HttpError, RetryError, RMPAPIError } from "./errors.js";
import { TokenBucket } from "./rateLimit.js";
/**
 * Low-level HTTP client with retries, rate limiting, and typed errors.
 * Typically you use {@link RMPClient}, which uses this internally.
 */
export class HttpClient {
    _config;
    _bucket;
    _abortController = null;
    constructor(config) {
        this._config = config;
        this._bucket = new TokenBucket(config.rate_limit_per_minute, config.rate_limit_per_minute / 60);
    }
    _headers(extra) {
        const headers = {
            ...this._config.default_headers,
            "User-Agent": this._config.user_agent,
        };
        if (extra)
            Object.assign(headers, extra);
        return headers;
    }
    _url(path) {
        if (path === "")
            return this._config.base_url;
        const base = this._config.base_url.replace(/\/$/, "");
        const p = path.replace(/^\//, "");
        return `${base}/${p}`;
    }
    /**
     * Sends a POST request with a JSON body. Respects rate limit, timeout, and
     * retries. Throws {@link RMPAPIError} if the response contains a GraphQL
     * `errors` key. Throws {@link HttpError} on non-2xx status. Throws
     * {@link RetryError} after exhausting all retries.
     *
     * @param path - Path appended to base_url, or "" to POST directly to base_url.
     * @param payload - Object serialised as the JSON body.
     * @param headers - Optional extra headers merged into the request.
     */
    async postJson(path, payload, headers) {
        const url = this._url(path);
        let lastError = null;
        for (let attempt = 0; attempt <= this._config.max_retries; attempt++) {
            await this._bucket.consume();
            this._abortController = new AbortController();
            const timeoutId = setTimeout(() => this._abortController?.abort(), this._config.timeout_seconds * 1000);
            try {
                const response = await fetch(url, {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                        ...this._headers(headers),
                    },
                    body: JSON.stringify(payload),
                    signal: this._abortController.signal,
                });
                clearTimeout(timeoutId);
                this._abortController = null;
                if (response.ok) {
                    const data = (await response.json());
                    if (data && typeof data === "object" && "errors" in data) {
                        throw new RMPAPIError("RMP API returned errors", data.errors);
                    }
                    return data;
                }
                const body = await response.text();
                const err = new HttpError(response.status, url, body);
                lastError = err;
                if (response.status >= 500 && response.status < 600 && attempt < this._config.max_retries) {
                    continue;
                }
                throw err;
            }
            catch (e) {
                clearTimeout(timeoutId);
                this._abortController = null;
                if (e instanceof HttpError || e instanceof RMPAPIError)
                    throw e;
                lastError = e instanceof Error ? e : new Error(String(e));
                if (attempt >= this._config.max_retries) {
                    throw new RetryError(lastError);
                }
            }
        }
        throw new RetryError(lastError ?? new Error("Unknown error"));
    }
    /**
     * Aborts any in-flight request. Safe to call multiple times.
     */
    close() {
        this._abortController?.abort();
    }
}
